[uart] Multiple ESP32 features and fixes

brambo123 opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->
This PR adds 3 new configuration options to the uart component for ESP32:
- `flow_control_pin` can be used for ~~hardware~~ driver level RS485 flow control. Reducing the need for blocking software flow control.
- `rx_full_threshold` determines how many bytes are needed te be received before data is available for processing.
- `rx_timeout` determines how long (in byte-lengths) the bus can be idle before data becomes available for processing.

In addition, in the IDF platform version, some code was placed in the setup() function instead of load_settings().
This made it impossible to change some parameters during runtime.


The current fixed `rx_full_threshold` seems to be 120.
Newer Arduino framework versions seems to switch to 1 at low baudrates.
The current `rx_timeout` seems to be 10 for ESP32-IDF and 2 for Arduino, although Arduino seems to have opted for 1 in newer versions.
A `rx_timeout` of 10 byte-lengths adds a minimum of about 10.5ms delay at 9600 bauds before data becomes available. Arduino's 2 seems a more suitable default value.


## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- https://github.com/esphome/feature-requests/issues/421

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5352

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
uart:
  flow_control_pin: GPIO22
  rx_full_threshold: 8
  rx_timeout: 1

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

brambo123 committed:
[uart] Add hardware flow control

brambo123 committed:
[uart] flow control - fix ESP32 Arduino < 2.0.8

brambo123 committed:
[uart] lint fix

brambo123 committed:
[uart] fix lint

brambo123 committed:
[uart] fix lint

brambo123 committed:
[uart] fix Arduino framework check

brambo123 committed:
[uart] Improve hardware flow control, add interrupt control, fix runtime changes

- Allow hardware flow control on older ESP32-arduino framework versions
- Add rx_full_threshold and rx_timeout parameters
- Move ESP32-IDF configuration from setup() to load_settings(), allowing runtime changes

brambo123 commented:
For information, the current fixed values for rx_timeout and rx_full_threshold:
[Arduino rx_timeout: 2](https://github.com/espressif/arduino-esp32/blob/2.0.5/cores/esp32/HardwareSerial.cpp#L143)
[Arduino rx_full_threshold: 112](https://github.com/espressif/arduino-esp32/blob/2.0.5/cores/esp32/HardwareSerial.h#L109)
[ESP-IDF rx_timeout: 10](https://github.com/espressif/esp-idf/blob/v5.1.5/components/driver/uart/uart.c#L56)
[ESP-IDF rx_full_threshold: 120](https://github.com/espressif/esp-idf/blob/v5.1.5/components/driver/uart/uart.c#L55)

brambo123 committed:
[uart] Move CONF_ variables

jesserockz commented on esphome/components/uart/__init__.py:
```diff
@@ -239,8 +243,17 @@ def validate_port(value):
+                cv.only_on_esp32, cv.validate_bytes, cv.uint8_t
+            ),
+            cv.SplitDefault(CONF_RX_TIMEOUT, esp32=2): cv.All(
+                cv.only_on_esp32, cv.validate_bytes, cv.int_range(min=0, max=92)
```
Shouldn't this be a time_period validator?

exciton replied:
No - this timeout is set in bytes. The actual time is calculated internally by the framework depending on the baud rate and accounting for start/stop/parity bits.

clydebarrow replied:
So maybe the option should be named differently?

exciton replied:
The naming matches the framework naming, so that's helpful, and it _is_ a timeout, just set in units of bytes. If the documentation is updated it shouldn't be confusing to people.

clydebarrow replied:
Step out of the developer mindset for a moment; the average user will know nothing about the framework and won't read the docs unless forced to. And it already confused Jesse :-)
IMO the framework function is confusingly named, there is a brief opportunity here to not repeat that mistake.

brambo123 replied:
It is just a timeout, just not with second scales.
Calling it rx_timeout_in_byte_lengths is also a bit weird.
A better name is welcome.

By the way, I expect that a average user only sets this value if a component indicates that a certain value is required  ;)

DjordjeMandic replied:
`rx_byteout` (:
`rx_timeout_bytes`

jesserockz replied:
Indeed, Its confusing to call something a **time**out when it has nothing to do with time.

It is not exactly bytes too, so I don't have a good alternative name

```c
 * @param tout_thresh  This parameter defines timeout threshold in uart symbol periods. The maximum value of threshold is 126.
 *        tout_thresh = 1, defines TOUT interrupt timeout equal to transmission time of one symbol (~11 bit) on current baudrate.
 *        If the time is expired the UART_RXFIFO_TOUT_INT interrupt is triggered. If tout_thresh == 0,
 *        the TOUT feature is disabled.
```

kbx81 replied:
Confusing naming -- I also agree.
If I'm understanding this correctly, the timeout is measured in intervals of something ("uart symbol periods").
Maybe `CONF_RX_TIMEOUT_INTERVALS` or `CONF_RX_TIMEOUT_INTERVAL_COUNT` would be better?

kbx81 replied:
Are we agreeing to stick with `CONF_RX_TIMEOUT` for the name here?

kbx81 replied:
Left some suggestions on the docs; in particular, [this suggestion](https://github.com/esphome/esphome-docs/pull/5352#discussion_r2356202639) is related to this discussion and I wanted to call it out for feedback. 🙂

exciton commented on esphome/components/uart/__init__.py:
```diff
@@ -239,8 +243,17 @@ def validate_port(value):
+            ),
             cv.Optional(CONF_PORT): cv.All(validate_port, cv.only_on(PLATFORM_HOST)),
             cv.Optional(CONF_RX_BUFFER_SIZE, default=256): cv.validate_bytes,
+            cv.SplitDefault(CONF_RX_FULL_THRESHOLD, esp32=120): cv.All(
```
Can we emulate the recent library behaviour of setting this to 120 for high bauds and 1 for <= 9600)?

brambo123 replied:
I was thinking about what a good default value is.
Something that doesn't cause an unnecessary amount of interrupts, but does provide good throughput.

What if we calculate a default value based on a maximum time?
For example limited to 10ms.
Then you have good throughput at any baudrate.

exciton replied:
That could work, but it would definitely be different from what the frameworks currently use, and may change behaviour for folks unexpectedly. 
It may be better to try to emulate what the frameworks do (as a default), while making it super clear to people what the default is (currently it's hidden deep in framework code). 
e.g. Maybe like this: https://github.com/espressif/arduino-esp32/blob/b63f947037bece7a499faf0043636dd8531b885e/cores/esp32/HardwareSerial.cpp#L414

brambo123 replied:
Only newer framework versions adjust the default value.
If I'm correct, the recommended Arduino version is 2.0.5, which has a fixed value.

The frameworks already use different default values.
And depending on the framework version there is also a difference.
For example, newer Arduino versions also use an rx_timeout of 1 instead of 2.

So what is the sensible thing to do?
I think it is wise to choose a default value for ESPHome that does not change across different frameworks and versions.
We must not forget that ESP32 is not the only platform within ESPHome.
Other platforms work differently, if a component relies on the high rx_full_threshold, it will not work on other platforms.
So I think it is safe to lower the default values.
Raising it later is not wise.

exciton replied:
Fair enough. Setting it to 120 / 2 seems good.

brambo123 replied:
I do think 120 is a bit high for lower baudrates.
Maybe other developers can also let us know what they think is wise.

Keeping the current value is nice.
But at a baudrate of 9600 a byte lasts about 1ms.
So with long data streams you would only get data every 120ms, a bit annoying with timeout detection.
And for general users it's a bit unfriendly to have to set a weird threshold all the time.

That's why I thought a default based on a safe 10ms would be useful.
Or we could go all the way to time in ms.

patagonaa replied:
Both approaches (setting the default to a fixed time based on the baud rate or setting it to 120 by default and to 2 when baud rate is low) sound okay, I think I personally would prefer a fixed time, so time-based timeouts (like there are in the modbus component) always work the same no matter the baudrate.

Either way, even if left in the current state, this should definitely be merged, as without this, modbus doesn't work properly (and can't be configured to) when the response is long and the baudrate is low.

brambo123 commented:
I came across a small thing.
The get_rx_full_threshold() (and other functions) should of course also indicate a good value on other platforms.
I will adjust it a bit.

And of course the tests need to be updated.
Which I think contains an error anyway.
Someone once thought that ESP32-S3 only has 2 uart ports, should be 3 as far as I know.

exciton commented:
I've already written code to work around any value that's set here for the
modbus component at least, so the default isn't super important. Simpler is
always better in my opinion.

On Thu, 23 Jan 2025, 1:48 pm brambo123, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In esphome/components/uart/__init__.py
> <https://github.com/esphome/esphome/pull/8103#discussion_r1927723748>:
>
> >              cv.Optional(CONF_PORT): cv.All(validate_port, cv.only_on(PLATFORM_HOST)),
>              cv.Optional(CONF_RX_BUFFER_SIZE, default=256): cv.validate_bytes,
> +            cv.SplitDefault(CONF_RX_FULL_THRESHOLD, esp32=120): cv.All(
>
> I do think 120 is a bit high for lower baudrates.
> Maybe other developers can also let us know what they think is wise.
>
> Keeping the current value is nice.
> But at a baudrate of 9600 a byte lasts about 1ms.
> So with long data streams you would only get data every 120ms, a bit
> annoying with timeout detection.
> And for general users it's a bit unfriendly to have to set a weird
> threshold all the time.
>
> That's why I thought a default based on a safe 10ms would be useful.
> Or we could go all the way to time in ms.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esphome/esphome/pull/8103#discussion_r1927723748>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALST5GI3LXHLGOEWJUT2WL2MFPSJAVCNFSM6AAAAABVMCRX2GVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNZRGA4TAMRQHA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

brambo123 commented:
I was thinking about this and then I thought of something.
Both rx_full_threshold and rx_timeout are purely for internal timing.
Shouldn't we just give components the freedom to adjust it?
And so users don't have to adjust it.

exciton commented:
Probably should let users set it, given the number of configurations that
are possible. Components can always ensure it's set correctly.




On Fri, 24 Jan 2025, 7:54 am brambo123, ***@***.***> wrote:

> I was thinking about this and then I thought of something.
> Both rx_full_threshold and rx_timeout are purely for internal timing.
> Shouldn't we just give components the freedom to adjust it?
> And so users don't have to adjust it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esphome/esphome/pull/8103#issuecomment-2612854463>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALST5HMUZVAWSWVZZQPDB32MJO2VAVCNFSM6AAAAABVMCRX2GVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJSHA2TINBWGM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

brambo123 committed:
[uart] Improve rx_full_threshold / rx_timeout control

- Add set_rx_full_threshold_ms()
- Make set parameters work without need to call load_settings()
- Allow more direct control for components

brambo123 committed:
[uart] Update tests

- Add new parameters to tests
- Fix ESP32-S3 max with usb tests

brambo123 committed:
[uart] Lint fix

exciton commented on esphome/components/uart/uart_component.cpp:
```diff
@@ -20,5 +20,19 @@ bool UARTComponent::check_read_timeout_(size_t len) {
+  uint8_t val = (this->baud_rate_ / (bytelength * 1000 / time)) - 1;
+  if (val <= 1) {
+    val = 1;
+    this->set_rx_timeout(0);
```
This disables the timeout interrupt, which may still be desired even if the full_threshold is 1. Better to let folks set this manually, since they may want to use the interrupt

brambo123 committed:
[uart] update set_rx_full_threshold_ms

- Change to int32_t to prevent overflow
- dont use set_rx_timeout

brambo123 committed:
[uart] allow set parameters before ready

exciton commented:
This looks good to me. Once merged I will update #8032 to utilize this. @jesserockz

patagonaa commented:
Please merge this and ideally set a good default for the rx_full_threshold (lower than common timeouts).

I just spent hours of my life staring at my logic analyzer and the modbus component code, wondering why it would send requests before the last response is received.

Turns out, of course, the receive timeout elapses because no bytes "arrive" (which would prevent the timeout from elapsing), allowing the next request to be sent.

kbx81 committed:
Merge branch 'dev' into uart_flow_control

kbx81 committed:
Whoops

kbx81 commented on esphome/components/uart/uart_component.cpp:
```diff
@@ -20,5 +20,18 @@ bool UARTComponent::check_read_timeout_(size_t len) {
+    val = 1;
+  } else if (val > 120) {
+    val = 120;
+  }
```
`#include "esphome/core/helpers.h"` and use `clamp` here:
```suggestion
  int32_t val = clamp<int32_t>((this->baud_rate_ / (bytelength * 1000 / time)) - 1, 1, 120);
```

brambo123 committed:
Merge branch 'esphome:dev' into uart_flow_control

exciton commented:
This looks ready to merge to me. When this is merged I can update #8032 to use this feature directly.

exciton commented on esphome/components/uart/uart_component_esp32_arduino.cpp:
```diff
@@ -161,6 +171,16 @@ void ESP32ArduinoUARTComponent::dump_config() {
 
+void ESP32ArduinoUARTComponent::set_rx_full_threshold(size_t rx_full_threshold) {
+  this->rx_full_threshold_ = rx_full_threshold;
+  this->hw_serial_->setRxTimeout(this->rx_full_threshold_);
```
this should be setRxFIFOFull

brambo123 committed:
Merge branch 'esphome:dev' into uart_flow_control

brambo123 committed:
Merge branch 'esphome:dev' into uart_flow_control

brambo123 committed:
Fix ESP32 Arduino

- Change ESP32 Arduino setMode function
- Fix ESP32 Arduino set_rx_full_threshold
- Use clamp() function

patagonaa reviewed (approved):
Code looks great and I've been running this for months now (to fix issues with long modbus responses) without any problems.

kbx81 commented on esphome/components/uart/uart_component_esp32_arduino.cpp:
```diff
@@ -151,8 +157,11 @@ void ESP32ArduinoUARTComponent::dump_config() {
   if (this->rx_pin_ != nullptr) {
     ESP_LOGCONFIG(TAG, "  RX Buffer Size: %u", this->rx_buffer_size_);
+    ESP_LOGCONFIG(TAG, "  RX Full Threshold: %u", this->rx_full_threshold_);
+    ESP_LOGCONFIG(TAG, "  RX Timeout: %u", this->rx_timeout_);
```
`ESP_LOGCONFIG` lines need to be consolidated into a single call. See our [logging best practices](https://developers.esphome.io/architecture/logging/#the-solution-combine-related-log-messages) (or look at any other components; they've all been updated accordingly).

patagonaa commented:
So, as far as I can tell, the open points are:
- default value for `rx_full_threshold` (I would propose 1 for baudrate <= 9600, 120 for everything else, as Arduino uses and should be good for most cases)
- naming for `rx_timeout` (maybe `rx_timeout_bytes`? Either way, a short description in the docs should clear up confusion that comes up)

@brambo123 are you planning to keep working on this PR as there hasn't been communication from you in a while?

To the ESPHome maintainers:
If brambo123 doesn't want to keep working on this, is there a way for me to complete this PR? I have a bunch of free time right now and am somewhat familiar with the ESPHome codebase.  
Also, what do you thing about my proposals above?

jesserockz commented:
@patagonaa Given the way GitHub works with PRs and branches/permissions etc, the best way for you to continue this PR would be to work on the branch in your own fork, then comment back here again and a maintainer is able to pull your commits into this branch. 

Either that, or @brambo123 needs to give you write access to their fork so you can push directly into the branch/PR.

brambo123 commented:
> > So, as far as I can tell, the open points are:
> > 
> > * default value for `rx_full_threshold` (I would propose 1 for baudrate <= 9600, 120 for everything else, as Arduino uses and should be good for most cases)
> 
> I support this. Will make for good defaults for modbus.

If we're going to use a different default value anyway...
Wouldn't it be better to go with a more accurately calculated value, like 10ms?
That way, you'll always have a reasonable threshold for any baudrate.
We can use the new set_rx_full_threshold_ms function for that.

brambo123 commented:
> * naming for `rx_timeout` (maybe `rx_timeout_bytes`? Either way, a short description in the docs should clear up confusion that comes up)

No one has a good alternative.
Adding _bytes doesn't automatically make it correct.
And then, in all other cases where time is involved, you'd also have to add things like _s or _ms.

> @brambo123 are you planning to keep working on this PR as there hasn't been communication from you in a while?
> 

Yes, I was just waiting for any reply

bigretromike commented:
@clydebarrow @jesserockz any chance to review this

brambo123 committed:
Merge branch 'esphome:dev' into uart_flow_control

brambo123 commented:
> Regarding `rx_timeout`:
> 
> > No one has a good alternative.
> > Adding _bytes doesn't automatically make it correct.
> 
> Maybe changing this timeout to actual time instead would be a good solution then (rounding up to 1 byte if it is below that), as I don't think it is ever really necessary to set a byte-exact timeout.

`rx_timeout` shouldn't be in time.
Changing this to time actually causes problems because the value becomes too high.
The default value of 2 is almost always fine.
Only in very specific cases may adjustment be desirable.

It is only the `rx_full_threshold` that is currently giving problems, and only for this value could you actually use time.
I feel like it might be better if the modbus component itself called set_rx_full_threshold_ms().
Creating a default value based on baud rate or time is a bit tricky to do in a neat way, mainly because the default is handled in the Python config verification.

brambo123 committed:
Change rx_full_threshold default

Set default value of rx_full_threshold to be calculated at 10ms

brambo123 committed:
Update dump_config

brambo123 committed:
Merge branch 'dev' into uart_flow_control

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

brambo123 committed:
Fix rx_full_threshold minimum value

Minimum for rx_full_threshold should be 1

exciton commented on esphome/components/uart/uart_component_esp32_arduino.cpp:
```diff
@@ -162,6 +173,16 @@ void ESP32ArduinoUARTComponent::dump_config() {
 
+void ESP32ArduinoUARTComponent::set_rx_full_threshold(size_t rx_full_threshold) {
+  this->rx_full_threshold_ = rx_full_threshold;
+  this->hw_serial_->setRxFIFOFull(this->rx_full_threshold_);
```
This line should be deleted, or changed to guard against null hw_serial_. As-is this causes a default configuration to crash at start-up because it's called before hw_serial_ is initialised.

exciton commented on esphome/components/uart/uart_component_esp32_arduino.cpp:
```diff
@@ -162,6 +173,16 @@ void ESP32ArduinoUARTComponent::dump_config() {
+
+void ESP32ArduinoUARTComponent::set_rx_timeout(size_t rx_timeout) {
+  this->rx_timeout_ = rx_timeout;
+  this->hw_serial_->setRxTimeout(this->rx_timeout_);
```
Likewise this needs to be deleted or guarded against a null hw_serial_.

exciton reviewed (changes requested):
Couple of changes to make this actually work with arduino framework. As is, it crashes on boot.

asund commented:
Thank you all for the effort. This is going to make communications with my 1200 baud device much quicker.

brambo123 committed:
Fix arduino framework

Only use this->hw_serial_ if component is_ready()

brambo123 commented:
> Couple of changes to make this actually work with arduino framework. As is, it crashes on boot.

Good find.
I think it worked in the past, but a newer Arduino framework has been adopted, so perhaps something has changed.
The ESP-IDF firmware already had a check for this, and the same check is now included in the Arduino Framework.

exciton reviewed (approved):
Tested with latest fix for arduino platform, and works well. Default rx_full_threshold of 10ms makes sense for esphome - this should ensure bytes are always available between component loops if any have been received. 
Testing was done along with #8032 (skipping the uart changes currently in that PR), and everything works together very nicely!

exciton commented:
@kbx81 - this PR is good to go, and implements all the UART features needed for #8032.

kbx81 commented:
Looks like a couple of merge conflicts have crept in 🙈

brambo123 committed:
Merge remote-tracking branch 'upstream/dev' into uart_flow_control

kbx81 reviewed (approved):
Thanks for working on this! 🍻

brambo123 commented:
> Is there anything we should do further testing on before merging?
Just to be sure, I did another test with the Arduino platform, works fine.
So I think it's ready to merge.

bdraco committed:
Merge branch 'dev' into uart_flow_control

bdraco commented:
clang-tidy failure looks transient. We can re-run once its done

kbx81 committed:
Merge branch 'dev' into uart_flow_control

kbx81 commented:
Installed on a few devices I have here which use the UART component, just for an additional smoke check...seems to be working fine. Thanks for all your work! 🍻

DavidAntonin commented:
With this PR and ESPHome 9.1 I am getting compile error:
src/esphome/components/uart/uart_component_esp_idf.cpp: In member function 'virtual void esphome::uart::IDFUARTComponent::setup()':
src/esphome/components/uart/uart_component_esp_idf.cpp:78:30: error: 'class esphome::logger::Logger' has no member named 'get_uart_num'; did you mean 'get_uart'?
   78 |       logger::global_logger->get_uart_num() == next_uart_num) {
      |                              ^~~~~~~~~~~~
      |                              get_uart
*** [.pioenvs/esp32-s2-bazen/src/esphome/components/uart/uart_component_esp_idf.cpp.o] Error 1
